hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsckViolationHandler.java

```diff
@@ -0,0 +1,315 @@
+        LOG.error("Can not instantiate handler: {}",
+            violation.getHandler());
+      }
+      sB.append(newLine);
+    }
+    LOG.error(sB.toString());
+  }
+
+  /**
+   * Violation handler abstract class.
+   * This class should be extended for violation handlers.
+   */
+  public static abstract class ViolationHandler {
+    private final PathMetadata pathMetadata;
+    private final S3AFileStatus s3FileStatus;
+    private final S3AFileStatus msFileStatus;
+    private final List<FileStatus> s3DirListing;
+    private final DirListingMetadata msDirListing;
+
+    public ViolationHandler(S3GuardFsck.ComparePair comparePair) {
+      pathMetadata = comparePair.getMsPathMetadata();
+      s3FileStatus = comparePair.getS3FileStatus();
+      if (pathMetadata != null) {
+        msFileStatus = pathMetadata.getFileStatus();
+      } else {
+        msFileStatus = null;
+      }
+      s3DirListing = comparePair.getS3DirListing();
+      msDirListing = comparePair.getMsDirListing();
+    }
+
+    abstract String getError();
+
+    public PathMetadata getPathMetadata() {
+      return pathMetadata;
+    }
+
+    public S3AFileStatus getS3FileStatus() {
+      return s3FileStatus;
+    }
+
+    public S3AFileStatus getMsFileStatus() {
+      return msFileStatus;
+    }
+
+    public List<FileStatus> getS3DirListing() {
+      return s3DirListing;
+    }
+
+    public DirListingMetadata getMsDirListing() {
+      return msDirListing;
+    }
+  }
+
+  /**
+   * The violation handler when there's no matching metadata entry in the MS.
+   */
+  public static class NoMetadataEntry extends ViolationHandler {
+
+    public NoMetadataEntry(S3GuardFsck.ComparePair comparePair) {
+      super(comparePair);
+    }
+
+    @Override
+    public String getError() {
+      return "No PathMetadata for this path in the MS.";
+    }
+  }
+
+  /**
+   * The violation handler when there's no parent entry.
+   */
+  public static class NoParentEntry extends ViolationHandler {
+
+    public NoParentEntry(S3GuardFsck.ComparePair comparePair) {
+      super(comparePair);
+    }
+
+    @Override
+    public String getError() {
+      return "Entry does not have a parent entry (not root)";
+    }
+  }
+
+  /**
+   * The violation handler when the parent of an entry is a file.
+   */
+  public static class ParentIsAFile extends ViolationHandler {
+
+    public ParentIsAFile(S3GuardFsck.ComparePair comparePair) {
+      super(comparePair);
+    }
+
+    @Override
+    public String getError() {
+      return "An entry’s parent is a file";
+    }
+  }
+
+  /**
+   * The violation handler when the parent of an entry is tombstoned.
+   */
+  public static class ParentTombstoned extends ViolationHandler {
+
+    public ParentTombstoned(S3GuardFsck.ComparePair comparePair) {
+      super(comparePair);
+    }
+
+    @Override
+    public String getError() {
+      return "The entry's parent tombstoned";
```

Comment:
again, better text. Which needs to get in the docs too. How about "The entry in the metastore database has a parent entry which is a tombstone marker"